concent_api/core/payments/report.py

```diff
@@ -0,0 +1,202 @@
+            subtask.requestor,
+            subtask.provider,
+        )
+        related_pending_responses_for_pair = get_related_pending_responses_for_subtasks_ids(subtasks_ids, subtask.requestor, clients_ethereum_addresses)
+        try:
+            related_payments_for_pair = get_related_payments_for_pair(clients_ethereum_addresses)
+        except SCINotSynchronized:
+            related_payments_for_pair = Payments([], [], [], [], [])
+        is_related_payments_for_pair_empty = related_payments_for_pair.is_empty()  # pylint: disable=no-member
+        is_related_pending_responses_for_pair_empty = related_pending_responses_for_pair.is_empty()  # pylint: disable=no-member
+        if not(is_related_payments_for_pair_empty and is_related_pending_responses_for_pair_empty):  # pylint: disable=no-member
+            pair_report = PairReport(
+                requestor=subtask.requestor,
+                provider=subtask.provider,
+                requestor_ethereum_addresses=[client_ethereum_address.requestor for client_ethereum_address in clients_ethereum_addresses],
+                provider_ethereum_addresses=[client_ethereum_address.provider for client_ethereum_address in clients_ethereum_addresses],
+                pending_responses=related_pending_responses_for_pair,
+                payments=related_payments_for_pair,
+            )
+            pair_reports.append(pair_report)
+
+    return Report(pair_reports)
+
+
+def get_subtasks_with_distinct_clients_pairs() -> QuerySet:
+    """ Get Subtasks with distinct pairs of requestor and provider Clients. """
+    return Subtask.objects.order_by('provider', 'requestor').distinct('provider', 'requestor')
+
+
+def get_subtask_ids_and_ethereum_addresses_for_clients_pair(requestor: Client, provider: Client) -> Tuple:
+    """ Get Subtasks Ids and all Ethereum keys for Clients pair. """
+    subtask_ids = []
+    clients_ethereum_addresses = []
+
+    for subtask in Subtask.objects.filter(requestor=requestor, provider=provider):  # pylint: disable=no-member
+        subtask_ids.append(subtask.subtask_id)
+        task_to_compute = deserialize_message(subtask.task_to_compute.data.tobytes())
+        clients_ethereum_addresses.append(
+            ClientEthereumAddresses(
+                task_to_compute.provider_ethereum_address,
+                task_to_compute.requestor_ethereum_address,
+            )
+        )
+
+    return subtask_ids, list(set(clients_ethereum_addresses))
+
+
+def get_related_pending_responses_for_subtasks_ids(subtasks_ids: List[str], requestor: Client, clients_ethereum_addresses: list) -> PendingResponses:
+    """ Group and return pending responses from Subtask queryset. """
+    return PendingResponses(
+        subtask_results_settled=PendingResponse.objects.filter(
+            response_type=PendingResponse.ResponseType.SubtaskResultsSettled.name,  # pylint: disable=no-member
+            subtask__subtask_id__in=subtasks_ids,
+        ).distinct('subtask'),
```

Comment:
Why do you need `distinct('subtask')` here?

Reply:
Because we don't want duplicated `PendingResponses` which are stored in database for requestor and provider. Thanks to this we got only one message

Reply:
According to https://docs.djangoproject.com/pl/2.1/ref/models/querysets/#distinct you should use `order_by('subtask')` too, but also change it to `subtask_id`.

But I am also worried about something else. If you for example have 4 PendingResponses of type `SubtaskResultsSettled` for single Subtask, then you would return only one of them. Is that ok?

Reply:
@kbeker @Jakub89 How was this case finally resolved? I see `.distinct('subtask')` is removed, so we are going to print all 1-4 messages?

Reply:
As discussed, we want to show all extra messages and mark them as warning, which will be done in next pull request related to payment report.